### ISSUE TITLE
PostgreSQL as alternate database provider

### ISSUE THREAD

**Issue by kspearrin** [bitwarden/server#453]
Adding another database provider should be fairly straight forward. Bitwarden's data access layer is abstracted away with repository interfaces which can be found under `src/Core/Repositories`. I've already done the work needed for making this possible. All that is left is re-writing all of the SQL tables, functions, stored procedures, etc over to PostgreSQL and then wiring them up repository implementations for PostgreSQL. I've started this work in the following commits, which can be followed as a model going forward:

https://github.com/bitwarden/server/commit/84800da1fb046ab2da0ad14f6096f9ac04f60405
https://github.com/bitwarden/server/commit/acef40ee82e81ac81d4b17de09a2c174a3b7bfbb

Choosing a database provider is as simple as either providing a connection string for MSSQL or PostgreSQL, whichever you prefer.

I do not have a lot of experience in working with PostgreSQL, so I will need some help to complete this task.

Questions:

- Are PostgreSQL functions equivalent to MSSQL stored procedures?

**Comment by jeremyVignelles** [comment]
I was just trying to setup bitwarden on my server and immediately saw ram usage getting high due to the use of SQL Server. I was wondering if there was a way to use another db provider, when I found this post.

I've never used PostgreSQL, but I'm curious to know if that would help my server support the extra load.

Why are you using stored procedures for? What other dbms-specfic feature are you using? Would SQLite be a candidate for small environments?

**Comment by tamaralo** [comment]
Hi there, long time lurker, first time poster.
As somebody who is highly interested in this feature I would like to help with at least my two cents.:

> Are PostgreSQL functions equivalent to MSSQL stored procedures?

PosgreSQL stored procedures capabilities surpass MSSQL's, as this was one of the key aspects of its development. For a short info/comparison:  [ https://stackoverflow.com/questions/339744/better-languages-than-sql-for-stored-procedures](url)

>I've never used PostgreSQL, but I'm curious to know if that would help my server support the extra load.

What would be your server's setup? Win or *nix? Usually I would say yes, it should reduce some of the load, but I don't want to give you some false hope. 

>I do not have a lot of experience in working with PostgreSQL, so I will need some help to complete this task.

I never helped I any open source project out of fear of doing something wrong, but if you guys need some help and are willing to help guide me on how to contribute I would gladly help.

**Comment by jeremyVignelles** [comment]
> I never helped I any open source project out of fear of doing something wrong, but if you guys need some help and are willing to help guide me on how to contribute I would gladly help.

You can't mess up a project on your own by doing bad things. If you have the right development skills, go ahead, the worst that could happen is spending too much time, but you'll always learn something in the process 😉

**Comment by kspearrin** [comment]
@tamaralo All database interaction in Bitwarden with MSSQL is done with stored procedures. You can find them (all all other schema) listed here: https://github.com/bitwarden/server/tree/master/src/Sql/dbo

My assumption is that we'd do the same in PostgreSQL using stored procedures (seems to behave differently in PostgreSQL) or functions. Our SQL data mapper (Dapper) seems to only work with PostgreSQL functions.

The majority of the work here is just translating the MSSQL schema (tables, indexes, sprocs/functions) over to their equivalent on PostgreSQL. After that it's just a matter of wiring them up to a C# repository class, which is fairly easy to do. Do you have enough experience in MSSQL and PostgreSQL to translate between the two? If so, I'd say you're qualified enough to help with this.

**Comment by tamaralo** [comment]
@jeremyVignelles  😊 I give it a try! I have a lot of fear/respect for opensource projects, as they seem to be so well organized and fast paced. ^^

@kspearrin:
I worked with PostgreSQL, MySQL and some old versions from MSSQL ( in the 2000's ) so I think I could take a crack at it and try it at my server. The thing is, I study alongside my work and ATMI have some finals so, if its ok, I will try to take a look at it by the end of next week.  👍

**Comment by kspearrin** [comment]
@tamaralo If you have any questions when you get started, you can find me in our chatroom here: https://gitter.im/bitwarden/Lobby

**Comment by tamaralo** [comment]
@kspearrin Thx, I will come online by the end of next week (sorry, forgot the word *next* in the last post! Thanks for the prompt reply.

**Comment by dasbaumwolltier** [comment]
I would like to ask if I could also try my luck (although on the MySQL part, because I have never worked with postgres)? (I have also never helped in an open source project out of the same fear as tamaralo :D)

**Comment by Kovah** [comment]
The 2GB RAM requirement for SQLServer is the only thing holding me back from using Bitwarden. Would love to see an implementation for Postgres or any other database provider. 💙

**Comment by indy-singh** [comment]
Would love to help Bitwarden support PostgreSQL; though I tend to avoid stored procs/functions and prefer to bring that logic up into the code.

**Comment by mrahbar** [comment]
👍 For supporting PostgreSQL. I'm a big fan of Bitwarden and a happy paying customer. The only reason of not using a self-hosted Bitwarden is MSSQL.

On the topic tough I think that without any official endorsement of the Bitwarden team for PostgreSQL this endeavor is be doomed to failure. After the initial batch of SQL scripts is ported who is going to create new PostgreSQL scripts every time the equivalent MSSQL scripts are updated? Kudos for a defined data-abstraction-layer but I'm not optimistic when it gets to maintain two different database vendors. 

So can this process be only an experiment or a real alternative solution for MSSQL?

**Comment by jonny3010** [comment]
👍
This sounds like a good idea to me. Getting ram usage down and removing minimum 2 GB limit is good. Will make running it more resource efficient on virtual machines etc...

Although we do need support from Bitwarden themselves like @mrahbar says.

**Comment by Agraphie** [comment]
I would like to see this as well. Are there any experimental branches or any way I can help make this happen?

**Comment by CaptAintHere** [comment]
Any news about this?

**Comment by devployment** [comment]
As briefly mentioned [here](https://gitter.im/bitwarden/Lobby?at=5d0777c0faf70031f92c5eb7) I'll try to volunteer here. Will set things up locally and see where I hit roadblocks.

**Comment by jeremyVignelles** [comment]
@devployment : 👏 and good luck.
Do you have any experience in working with Postgre SQL?
Please keep us posted 🙂

**Comment by devployment** [comment]
> Do you have any experience in working with Postgre SQL?

@jeremyVignelles, yes I do. It's been a while. But we'll see.

**Comment by benjaminpreiss** [comment]
Is anyone already working on translating the schemes?

**Comment by kspearrin** [comment]
Not that I am aware of.

**Comment by wusatosi** [comment]
> > Do you have any experience in working with Postgre SQL?
> 
> @jeremyVignelles, yes I do. It's been a while. But we'll see.

Any update?

**Comment by benjaminpreiss** [comment]
I have started with translating but made close to no progress.

**Comment by Papina** [comment]
Just read this while investigating postgreSQL. I have good experience with tsql to pl/pgsql so I might start to give this a crack.

**Comment by Papina** [comment]
im down to the nitty gritty now of functions. Im not super familiar with docker, but i cant seem to publish the mssql ports from the docker image successfully. Is there a switch i can add that will expose the 1433 port?  when i add it it i get an error os OSX

i modified the function to publish port 1433:


```
function updateDatabase() {
    pullSetup
    docker run -i --rm --name setup -p 1433:1433 --network container:bitwarden-mssql \
        -v $OUTPUT_DIR:/bitwarden --env-file $ENV_DIR/uid.env bitwarden/setup:$COREVERSION \
        dotnet Setup.dll -update 1 -db 1 -os $OS -corev $COREVERSION -webv $WEBVERSION
    echo "Database update complete"
}
```

now i get the error:

```
MacBook-Pro:bitwarden benjamin$ ./bitwarden.sh updatedb
 _     _ _                         _
| |__ (_) |___      ____ _ _ __ __| | ___ _ __
| '_ \| | __\ \ /\ / / _` | '__/ _` |/ _ \ '_ \
| |_) | | |_ \ V  V / (_| | | | (_| |  __/ | | |
|_.__/|_|\__| \_/\_/ \__,_|_|  \__,_|\___|_| |_|

Open source password management solutions
Copyright 2015-2019, 8bit Solutions LLC
https://bitwarden.com, https://github.com/bitwarden

===================================================

Docker version 19.03.2, build 6a30dfc
docker-compose version 1.24.1, build 4667896b

1.32.0: Pulling from bitwarden/setup
Digest: sha256:e88f1611ff88c77a6255c49189ac3c965aaa3576fa6980ba54f2be10a96907b5
Status: Image is up to date for bitwarden/setup:1.32.0
docker.io/bitwarden/setup:1.32.0
docker: Error response from daemon: conflicting options: port publishing and the container type network mode.
See 'docker run --help'.
```

this is more for testing, so i can make sure the functions im rewriting actually produce the same data

**Comment by kspearrin** [comment]
You want to connect to the MSSQL instance of a Bitwarden installation? Just bash into the mssql docker container. Then you can run `sqlcmd`.

`$ docker exec bitwarden-mssql bash`

**Comment by Papina** [comment]
no, i want to connect to the mssql instance from the host, so i can run the development tools against the database directly, but still be able to use the bitwarden normally at the same time.

**Comment by Papina** [comment]
![Screen Shot 2019-09-20 at 10 46 47](https://user-images.githubusercontent.com/11512841/65290809-f8a8d380-db93-11e9-9672-6f1a85daba43.png)

**Comment by kspearrin** [comment]
Ok. Then create the following file `./bwdata/docker/docker-compose.override.yml`.

```
version: '3'

services:
  mssql:
    ports:
      - '1433:1433'
```

Restart. Now you can connect to `localhost:1433` from tooling on the host.

**Comment by Papina** [comment]
awesome! got it working finally in dbeaver. 

screenshot for those attempting the same, it needs a non-obvious driver, not the driver that is normally used for a windows SQL instance

![Screen Shot 2019-09-20 at 12 28 10](https://user-images.githubusercontent.com/11512841/65294732-71169100-dba2-11e9-9ec9-923b7c35da5c.png)

I will continue with the conversions.

**Comment by knocte** [comment]
@Papina glad to see some progress; instead of asking for an update I'm going to chip in some funds with gitcoin. Hopefully other people interested in this can join me (adding more funds to it) to incentivise this work to be finished and merged to master.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.276 ETH (50.03 USD @ $181.26/ETH)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/bitwarden/server/453/3550).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/bitwarden/server/453/3550).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $155,958.85 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 2 months, 2 weeks from now.
Please review their action plans below:


**1) [papina](https://gitcoin.co/profile/papina) has started work.**

1funtion per week until complete

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/bitwarden/server/453/3550).

**Comment by Papina** [comment]
Just started a new job, so I have been busy, but I have a few functions translated, just haven't pushed the updates

**Comment by FingerlessGlov3s** [comment]
Whats the latest update on the conversion?

**Comment by Papina** [comment]
waiting on two pull requests before i continue further

https://github.com/bitwarden/server/pull/578
https://github.com/bitwarden/server/pull/580

**Comment by knocte** [comment]
@Papina PR #578 still has a typo which the maintainer is requesting to be addressed. And #580 needs to be rebased because it has conflicts. So it looks like what needs to be done is take action and not wait, or am I missing something?

**Comment by Papina** [comment]
spelling mistake is fixed, didnt see that review. 
the conflicts are updates to existing views, just adding in 'OR REPLACE' to various DDL

**Comment by knocte** [comment]
There are now conflicts in both PRs :)

**Comment by Morl99** [comment]
Hello,
I want to chime in and share my opinion that the current approach is futile and will leave the project in a state where maintenance is really hard. 

I suggest that we introduce Liquibase to manage the DBMS Setup. It should also support writing DBMS agnostic Functions/Procedures (never used it for that though). And the best part is that converting to Liquibase can be done tool based, see their documentation: https://www.liquibase.org/documentation/existing_project.html

I am willing to help with this, if the community and maintainers thinks that this is a good idea.

I am personally motivated because I want to run my own instance of Bitwarden, but mssql is a deal-breaker for my setup. I do have a lot of pssql experience (we run about 100 pssql instances at work, though we spin them up in AWS).

**Comment by knocte** [comment]
I'm no maintainer, but seeing the other approach seems to be getting no love, I'd go for it @Morl99, worst thing that can happen is that both strategies get committed and the one that has the least maintenance  work survives.

**Comment by Morl99** [comment]
I thought about it some more, and I fear that the approach is a good idea, but the tool might not be a good fit for a .NET environment, and it would be some work to bootstrap this at runtime (like adding a Java runtime to the docker container). I will look into similar tools for .NET. Or does anybody have any experience to share?

**Comment by mnuccioarpae** [comment]
I fear that there's no such thing as DBMS agnostic functions or procedures.

**Comment by Morl99** [comment]
Yeah I arrived at the same conclusion. But it would be a help if the schema itself would be declared in a DBMS agnostic way.

Another option would be to reduce the amount of procedures and pull some of that functionality into the application (using sql statements if necessary). 

I am specifically looking for recommendations for a dbms agnostic schema creation/update tool that plays well in a dot.net environment.

**Comment by knocte** [comment]
> Another option would be to reduce the amount of procedures and pull some of that functionality into the application (using sql statements if necessary).

I bet the best approach for that would be using EntityFramework or some other ORM, by I also bet that the maintainer of bitwarden would not be happy about this drastic change.

**Comment by Morl99** [comment]
Let us just ask 😉
@kspearrin do you have an objection against EntityFramework? I had used it in my Bachelors Thesis, and tough the development flow was really nice. But it is a long time ago and the project was a lot smaller than Bitwarden. It would surely be  challenge to convert, but I think it could also be a lot of fun.

**Comment by kspearrin** [comment]
@Morl99 Yes, I am not a fan of ORMs like EF. Too much “auto magic” code generation. Stored procedures are always preferred.

**Comment by kspearrin** [comment]
@Morl99 we could still consider EF as the base for other alternative database repository implementations, however. It’s just that we cannot use EF for our Sql server production environment which has a much larger scale than the on-prem installations people here are considering using Bitwarden for. So I’m the end we could have sql server repositories and EF repository implementations.

**Comment by jeremyVignelles** [comment]
In my experience, EF slows down the startup of application because of the huge reflection usage. Is that still true in 2019?

Yes, using EF could save a lot of dev time in providers implementation, but it adds a dependency that might not be wanted, and adds an overhead to the application.

**Comment by jeremyVignelles** [comment]
May I suggest the use of [Dapper](https://github.com/StackExchange/Dapper)?
It has been created with perf in mind at stackoverflow, and it seems database-agnostic.

**Comment by knocte** [comment]
IMO slow startup is not an issue for server apps because it just start once, and is running most of the time. Also IME Dapper is worse than EF (more weakly typed) but I don't want to start a flamewar here.

**Comment by jeremyVignelles** [comment]
It's not about starting a flamewar, we're just discussing alternatives. I don't have any decision power either.

After posting, I realized that Dapper might not be a good fit because you still need to write your SQL queries for anything more complex than a SELECT, so you're going to write DB-specific queries anyway.

You may be right, EF startup time might not be an issue for server-side applications, and I'm pretty sure that things got better since EF core.

**Comment by mnuccioarpae** [comment]
I agree that stored procedures are better. The problem I have is with the comments in the stored procedures, which do not explain the code.

For example in https://github.com/bitwarden/server/blob/master/src/Sql/dbo/Stored%20Procedures/User_DeleteById.sql there's no comment explaining why ciphers are deleted in batches of 100 records and why this is done in a separate transaction. What happens if the second transaction fails? The user remains, but his ciphers are gone? If yes, why this is desired behaviour?

I am pretty sure that all the T-SQL procedures can be translated in an equivalent PostgreSQL procedure (expecially if we require latest PostgreSQL), but it's hard to do without some help.

Maybe an automated tool like one of the ones listed at https://wiki.postgresql.org/wiki/Microsoft_SQL_Server_to_PostgreSQL_Migration_by_Ian_Harding can be helpful at least for the first translation?

**Comment by knocte** [comment]
> @Morl99 Yes, I am not a fan of ORMs like EF.

Another ORMless option (and also storeprocedure-less) would be using https://github.com/rspeele/Rezoom.SQL, the only downside is that this db layer would need to be written in F# instead of C#.

**Comment by kspearrin** [comment]
Now exploring Entity Framework for this. Started some repository implementations for it here:

https://github.com/bitwarden/server/commit/7c5be176faa5c83b1545d4fd96059f4f4a3b8426

If anyone is experienced with Entity Framework, I'd love for some help.

**Comment by mnuccioarpae** [comment]
Has anyone tried https://github.com/bigsql/pgtsql ?

**Comment by insub** [comment]
Is it done ? How to switch to PostgreSQL?

**Comment by Papina** [comment]
The basic translation of t-sql to postgres functions is done for 95% of the procs. However, there is a lot of functions to go through and renaming of columns etc. A few functions didn't translate and I will have to manually do these few

**Comment by knocte** [comment]
Renaming of columns, why?

**Comment by Papina** [comment]
You'll have to read all the commits, but the decision was made to keep to postgresql standards and use snake_case_formatting. So all the functions will have to be modified to match the new schema definitions which uses this format. A bit of extra work, but worth it to get it right the first time. Also AWS DMS wasn't perfect and so I'll need to clean up some of the converted code. Most of it search/replace . But a few functions will need to be rewritten where the conversion just didn't know what to do. Feel free to help! There is plenty of functions to clean up, but we are 80% there as far as stored functions goes IMHO

**Comment by ataraxus** [comment]
also longtime lurker here. backing off hosting bitwarden because of mssql. Is this the right place to get updates regarding implementaion of psql support?

**Comment by cyphase** [comment]
Just another interested potential user, but yes, I would say it is.

**Comment by knocte** [comment]
Guys if you're very interested in this, please chip in a small contribution in [the gitcoin issue](https://gitcoin.co/issue/bitwarden/server/453/3550). +1 comments are just noise.

**Comment by emericv** [comment]
Hi there!

For people want to have a lighter alternative than the official MSSQL based solution, you can consider a very light and compatible backend alternative : https://github.com/dani-garcia/bitwarden_rs
in addition, all the services including web view are running in a single docker instance with a memory print around 5 MB!

**Comment by MicWit** [comment]
> Now exploring Entity Framework for this. Started some repository implementations for it here:
> 
> [7c5be17](https://github.com/bitwarden/server/commit/7c5be176faa5c83b1545d4fd96059f4f4a3b8426)
> 
> If anyone is experienced with Entity Framework, I'd love for some help.

I have used entity framework in the past, but most of my career has been in PHP/MySQL,Oracle or directly in MSSQL. I don't have any spare time at the moment, but I am keen to try get Bitwarden running on MariaDB in order to utilise our Galera cluster (we have just started experimenting this and have multi master with encryption running and are about to move that onto WAN and go live).

Have you had much help with Entity Framework? Is there any documentation on the code etc (or things like relationship diagrams etc) for Bitwarden, or is it all in the code?

**Comment by TheRealTripleB** [comment]
Hi, what's the current status?

**Comment by knocte** [comment]
The current status is that I'm really saddened that no one has contributed even 10USD for this to happen. You guys like opensource projects? Then support them with your money. There are 48 upvotes on this issue, if each one of them contributed just 10USD, we would add 480USD to the funding, which would mean that we have a bounty higher than 500 USD, attractive enough for a competent developer to work on it.

**Comment by joekerna** [comment]
I want to contribute money to this cause. But where can I do it? I'm quite new to this...

**Comment by FingerlessGlov3s** [comment]
I've clicked the link https://gitcoin.co/issue/bitwarden/server/453/3550
signed in and can't see a way to add funding :-(

**Comment by ronnicek** [comment]
You will need to create account on some crypto exchange.. proceed with KYC then load money there buy ETH, create wallet for ETH, move money there and then connect it on gitcoin.. because why do it easy way when we can do it hard way?

**Comment by FingerlessGlov3s** [comment]
no wonder why only one person has backed it then.
That's hella a process!
Can't I just Paypal the PR requester when I happens 🤣

**Comment by knocte** [comment]
> proceed with KYC then load money there buy ETH

There are platforms that don't require KYC where you can buy crypto, e.g. LocalCryptos or HodlHodl.
Plus, there's no need to have ETH, you can use DAI (a stablecoin pegged to USD) with GitCoin.

> Can't I just Paypal the PR requester when it happens

PayPal can act as an intermediary to know if a PR has been done/merged or not. Plus, PayPal can freeze your money in case they are not happy with your opensource project or feature. This is where crypto comes in: no counterparty risk.

**Comment by TheRealTripleB** [comment]
Okay, well, I'm gonna have to figure out how to set up the money today. I would also appreciate it if other interested people would do the same.

**Comment by FingerlessGlov3s** [comment]
Yes I know Paypal can't be used for opensource code donations, its not the right platform, it was a joke more than anything 😉

I'll have a look later I do have some Eth in my Ledger but its investment hehe. So would need to go buymore from like coinbase then send it over to gitcoin if they have eth address for my account 🤔. Shall have a better look later.

**Comment by tomdavidson** [comment]
Ill pitch in too. My family is a paid customer. My company is a paid customer. I will also pitch in directly to this issue and have been evaluating the Rust implementation of Bitwarden over the db issue.

**Comment by shortspider** [comment]
Is this still looking to be done? Is all thats involved to get parity between `src/Core/Repositories/SqlServer` and `src/Core/Repositories/PostgreSql` and the associated functions?

**Comment by kspearrin** [comment]
We're now looking at using Entity Framework to implement alternate database engines. This is currently on our backlog.

**Comment by am-oxipay** [comment]
I'm willing to help here. I've ported many projects at work to postgresql. 
Is the desire to go to EF directly now ? If so can I have some guidance as to what I can safely work on without clashing with others.

**Comment by Spacelord09** [comment]
Hey guys! Any updates to this Topic?

**Comment by cscharf** [comment]
This is on our product roadmap for work by our team in 2021.

**Comment by GieltjE** [comment]
I agree with @MicWit MariaDB would be great as it supports master-master, making scaling much easier.

**Comment by cscharf** [comment]
Thanks @GieltjE , our approach will support multiple RDMBS including mySQL and PostgreSQL based engines (so could support MariaDB)

**Comment by cscharf** [comment]
FYI, we'll be starting work on this capability this week hopefully!

**Comment by nbebout** [comment]
@cscharf will there be a migration path for moving from MSSQL to postgres or mariadb/mysql?

**Comment by cscharf** [comment]
> will there be a migration path for moving from MSSQL to postgres or mariadb/mysql?

We will first ensure we have a process and migration path for version to version on each platform independently. Cross-platform migration may need to be something we do externally via script(s), utilities, etc. Not sure we'll be building his out initially or have capacity to but that's a good question. If you're using this personally or family worst case you may need to export JSON vault data, build a new environment, import that data or something to that degree but we'll see as we near completion on the work where we're at capacity-wise as a team on this (or if anyone in the community is interested in creating data transfer scripts/utilities for the same since it'll just be pretty basic table data).

**Comment by GieltjE** [comment]
If I'm reading the git changes correctly technicaly we shouldn't be far off, what are the plans for releasing it all?

**Comment by cscharf** [comment]
So this is all code-complete and Addison has done some extensive unit testing on both PostgreSQL and mySQL. Here's the rub, we don't have the QA resources at the moment to do a full-on in-depth regression suite across this at the moment, but will work at it this year. Also, our DevOps team is working on a manageable self-hosted deployment strategy (alternative docker containers for myssql or pgsql vs. mssql) which could take a little time. If anyone would like to _alpha_ test this thing, technically it's ready to go in the `master` branch.

**Comment by GieltjE** [comment]
Would it be possible to run it with an existing mysql server instead of the predelivered docker?

**Comment by douglasparker** [comment]
> So this is all code-complete and Addison has done some extensive unit testing on both PostgreSQL and mySQL. Here's the rub, we don't have the QA resources at the moment to do a full-on in-depth regression suite across this at the moment, but will work at it this year. Also, our DevOps team is working on a manageable self-hosted deployment strategy (alternative docker containers for myssql or pgsql vs. mssql) which could take a little time. If anyone would like to _alpha_ test this thing, technically it's ready to go in the `master` branch.

This is huge! I am looking forward to a stable release.

**Comment by cscharf** [comment]
> Would it be possible to run it with an existing mysql server instead of the predelivered docker?

Yes, absolutely! EF has generated the upgrade scripts so you should be able to run those manually against your mySQL database.

**Comment by TheRealTripleB** [comment]
that sounds very good, I will set up a test instance in the next few days

**Comment by GieltjE** [comment]
According to git (https://github.com/bitwarden/server/compare/v1.41.6...v1.42.0) the changes have been merged and released, is that correct?

**Comment by MGibson1** [comment]
_Kind_ of. We release our to our cloud instances prior to pushing changes to self-hosted. It's possible to build and run the currently released code, but the self-hosted version is **not** yet released so I can't guarantee there won't be a flurry of changes you'll need to cope with.

**Comment by GieltjE** [comment]
Is there a (quick) writeup/pointer somewhere of the required migration steps?

**Comment by GieltjE** [comment]
> 
> 
> > Is there an ETA on when we can use an alternative db provider for self hosted instances?
> 
> The code has been released, however it's still technically `alpha` as our QA team has not had a chance to test it fully; it does work and function and I believe a few folks have started using the bits for it already. You are more than welcome to start using it and reporting any issues you may come across or find. I would recommend against using it for your primary vault (w/o a solid backup strategy). Thanks!

If there is a accesable way to use it without the mysql/mariadb docker container (e.g. with an existing mariadb server) I'll be more than happy to try it.

**Comment by Gohph3l** [comment]
@tgreer-bw @MGibson1 do you have some updates for us? :)

Seems that the alternative DB backups feature [was requested since 2018](https://community.bitwarden.com/t/use-other-database-than-mssql/612) and was a "high priority" backlog item since at least 2020. But it never made it into the official roadmap even the current one, although it is almost finished?

Can we expect a fully implementation and documentation in the next months?

**Comment by FingerlessGlov3s** [comment]
I too would like an update on the where we are, as MSSQL is rather heavy for small user base 😅

**Comment by GieltjE** [comment]
Given the number of open [job positions](https://bitwarden.com/careers/), I don't think we're gonna see this anytime soon.

If there was a (even somewhat hacky) way to beta test this I would be more than happy to help find bugs.

**Comment by cscharf** [comment]
I am very sorry everyone, after seeing some of the recent posts I realized we've not shared our documentation on how to get all this setup and going for those wishing to help us test it out and report bugs. It is much appreciated and below you'll find what we have in our internal docs the team has been using. Please let me know if you find any issues or deficiencies in the documentation as well.

# Entity Framework (EF)

## Required Software

- Docker
- A way to manage user secrets in the server project (`dotnet` cli or VSCode with an extension)
- pgAdmin4 or your postgres database management software of choice

## Instructions

- Run `docker compose --profile postgres up`
- Open up pgAdmin4 and connect to your running database.
- Run the migration scripts saved in `server/util/PostgresMigrations/Scripts` in order from oldest to newest.
   - Alternatively, you can run `dotnet ef database update` and it will run all the migrations, however this is not as deterministic as using the migration scripts and should not be used in production environments.
- Add the following values to your API and Identity user secrets, changing information like root password as needed.
   - If you already have a Postgres database set up please adjust the value of your existing `databaseProvider` secret instead of creating a new one.

```json
"globalSettings:databaseProvider": "postgres",
"globalSettings:postgreSql:connectionString": "Host=localhost;Username=postgres;Password=example;Database=vault_dev;Include Error Detail=true",
```

- Open a terminal in the root of your `server` project and run `dotnet test`. Validate that all tests pass to verify your application is properly connected to the right database. If the tests pass, you should be okay, but consider checking through database tables in pgAdmin4 to make sure data was created.
 - You may need to also set up a MySql database for the tests to pass.

## Setting up a local vault using MySql and EntityFramework

- Run `docker compose --profile mysql up`

```json
"globalSettings:databaseProvider": "mysql",
"globalSettings:mySql:connectionString": "server=localhost;uid=root;pwd=example;database=vault_dev",
```

## F.A.Q

> Do I need to remove or modify my sqlServer connection string user secrets to connect to an Entity Framework based provider?

No. Connection strings from all possible database providers are meant to be stored in your secrets at all times. The only secret relevant to which provider you are targeting is `databaseProvider`, which defaults to MsSql and can be overridden to `'postgres'` or `'mysql'` to target those providers.

> Who generates migrations?

Developers before a PR is created.

> Who needs to run migrations scripts?

Anyone who wants to connect to an alternative database provider. The workflow here is the same as with the normal MsSql implementation: run the scripts in the `scripts` folders against their relating databases in chronological order.

# Generating Entity Framework Migrations

## Required Software

- A terminal
- The `dotnet` cli
- The `dotnet` cli Entity Framework Core tool [found here](https://docs.microsoft.com/en-us/ef/core/cli/dotnet)

## Instructions

- Ensure your user secrets in `API` have the correct provider configured in the `databaseProvider` field. The value should match the provider you are generating migrations for.
- Open a terminal in the root of the project you want to generate a migration for (either `util/PostgresMigrations` or `util/MySqlMigrations`)
- Run `dotnet ef migrations add [NAME_OF_MIGRATION]` to generate a migration
- Run `dotnet ef migrations script [LATEST_PREVIOUS_MIGRATION]` to output the SQL to be run against your database and run it.
- Save the script output in the `Scripts`folder of whichever project it is for.

**Comment by GieltjE** [comment]
@cscharf Could we run this against a non docker local postgresql/mysql too?

**Comment by kspearrin** [comment]
@GieltjE Yes, just adjust the connection string accordingly.

**Comment by GieltjE** [comment]
I'm strugling a bit with the instructions, some questions:

- If I'm reading the sources right, the migrations directory produces the same end result as running all the .sql/.psql files?
- What step/command actually copies all data from the mssql to the new db?
- We don't need the sources and can just use the available docker images?
- How/where do we set the secrets for the connection string and database provider?

**Comment by MGibson1** [comment]
Hi all, I figured some people might be interested in updating their self-hosted instance to use EF as well. I've put together some instructions on getting that up and running below. It's written to run postgres, but it should be fairly clear how to update to use MySql.

# Use Entity Framework on Self Hosted Instance

## Warnings
Entity Framework is a largely untested way to provide database services for Bitwarden. If you decide to use it, please expect inconsistencies and bugs. If you encounter any, please create an issue at  with as much reproduction information as possible.

We don’t yet have a way to automatically run database migrations. You need to do it yourself every time you update your Bitwarden instance. It’s best not to run a given migration multiple times. Again, this is not automated, so you’ll need to keep track yourself.

No attempt is made to migrate data from an existing MSSql database to a new database provider. You can swap back to MSSql at any time by changing `globalSettings__databaseProvider="postgres"` to `globalSettings__databaseProvider="sqlServer"`

DO NOT in a production instance with data you care about!

## Required Software
* Docker

## Instructions
* Install a Bitwarden self-hosted instance as normal: https://bitwarden.com/help/install-on-premise-linux/
  > ⚠️ These instructions have only been tested on linux
* Update `bwdata/docker/docker-compose.override.yml`
  > ⚠️ You should choose a better PgAdmin4 password

<details>
<summary>docker-compose.override.yml</summary>

``` yml
#
# Useful references:
# https://docs.docker.com/compose/compose-file/
# https://docs.docker.com/compose/reference/overview/#use--f-to-specify-name-and-path-of-one-or-more-compose-files
# https://docs.docker.com/compose/reference/envvars/
#
#########################################################################
# WARNING: This file is generated. Do not make changes to this file.    #
# They will be overwritten on update. If you want to make additions to  #
# this file, you can create a `docker-compose.override.yml` file in the #
# same directory and it will be merged into this file at runtime. You   #
# can also manage various settings used in this file from the           #
# ./bwdata/config.yml file for your installation.                       #
#########################################################################

version: '3'

services:
  mssql:

  web:

  attachments:

  api:

  identity:

  sso:

  admin:

  icons:

  notifications:

  events:

  nginx:

  postgresql:
    image: postgres:13.2
    container_name: bitwarden-postgres
    restart: always
    stop_grace_period: 60s
    networks:
      - default
      - public
    ports:
      - '5432:5432'
    volumes:
      - ../postgres/data:/var/lib/postgresql/data
      - ../postgres/config:/etc/postgresql
      - ../postgres/log:/var/log/postgresql
    env_file:
      - postgresql.env
      - ../env/uid.env
      - ../env/postgresql.override.env
  
  pgadmin:
    image: dpage/pgadmin4
    environment:
      PGADMIN_DEFAULT_EMAIL: postgres@bitwarden.com
      PGADMIN_DEFAULT_PASSWORD: root
    ports:
      - "8889:80"
    networks:
      - default
      - public
```
</details>

* Create `bwdata/docker/postgresql.env`
<details>
<summary>postgresql.env</summary>

``` env
POSTGRES_DB=bw_vault
POSTGRES_USER=postgres
```
</details>

* Create `bwdata/env/postgresql.override.env`

<details>
<summary>postgresql.override.env</summary>

``` env
POSTGRES_PASSWORD=example
```
⚠️ you should pick a better password
</details>

* Update `bwdata/env/global.override.env`
  * Add the following
  > ⚠️ be sure to update `Password` to match what you changed `example` to in `postgresql.override.env`

<details>
<summary>global.override.env</summary>

``` env
globalSettings__databaseProvider="postgres"
globalSettings__postgreSql__connectionString="Host=bitwarden-postgres;Port=5432;Username=postgres;Password=example;Database=bw_vault"
```
</details>

* Restart your bitwarden environment to apply these config changes: `bitwarden.sh restart`
* Run database migrations
  * Since we’re just setting up, we need to run all migrations. Clone the server repository `git clone https://github.com/bitwarden/server`
  * `cd` into the appropriate Scripts directory. For Postgres, it's `server/util/PostgresMigrations/Scripts`
  * Copy each migration to the postgres docker container and run it. The following script will do this for every migration in the current directory.

<details>
<summary>migrate everything script</summary>

``` bash
for f in `ls -v ./*.psql`; do docker cp $f bitwarden-postgres:/var/lib/postgresql/; docker exec -u 1001 bitwarden-postgres psql bw_vault postgres -f /var/lib/postgresql/$f; done;
```

This script needs to be run from the current directory, otherwise the $f in the docker exec command will not point to the script file on the container.
</details>

* Run bitwarden.sh restart just to make sure everything is up to date

Your self-hosted instance should now be using postgres!

**Comment by MGibson1** [comment]
@GieltjE The instructions Chad posted are what we use internally to set up development environments. For a self-hosted instance you'll need something l like what I just posted.

We don't yet have a solution for migrating data from SqlServer to another database provider. You'd be starting over from scratch. I really don't recommend doing this for an instance you rely on, anyway.

**Comment by cscharf** [comment]
Thanks @MGibson1 , and @Hinton also pointed out to me that I should strongly note that this should ***not*** be used in Production... just don't, PLEASE.

**Comment by holow29** [comment]
I tried to test this on a new install using MySQL (MariaDB) instance. I wanted to avoid MSSQL altogether. It appears that bitwarden-admin still requires MSSQL to run, at least at first while it tries to do migrations; without MSSQL, bitwarden-admin errors out after attempting too many failed connections. As it is now, it appears that MSSQL is either required at the start for migrations or required all the time for bitwarden-admin; I'm not sure because I was testing on a low-memory device and could not get MSSQL to run. Is there a plan to allow other databases right from install/without requiring MSSQL at all?

**Comment by holow29** [comment]
<details>
<summary>Another issue I've run into is with API server. It doesn't seem to be able to manage Collections properly because a query can't be translated: (expand)</summary>
<pre><code>[Error] The LINQ expression 'DbSet<CollectionCipher>()
    .Where(cc => cc.CipherId == __cipherId_0)
    .Select(cc => new { 
        CollectionId = cc.CollectionId, 
        CipherId = cc.CipherId
     })
    .LeftJoin(
        inner: __p_1
            .AsQueryable(), 
        outerKeySelector: t => t.CollectionId, 
        innerKeySelector: s => s.CollectionId, 
        resultSelector: (t, s) => new { 
            <>h__TransparentIdentifier0 = new { 
                t = t, 
                s_g = s_g
             }, 
            s = s
         })' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryTranslationPreprocessor.Process(Expression query)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryTranslationPreprocessor.Process(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at Microsoft.EntityFrameworkCore.DbContext.AddRangeAsync(IEnumerable`1 entities, CancellationToken cancellationToken)
   at Bit.Infrastructure.EntityFramework.Repositories.CollectionCipherRepository.UpdateCollectionsForAdminAsync(Guid cipherId, Guid organizationId, IEnumerable`1 collectionIds) in /home/runner/work/server/server/src/Infrastructure.EntityFramework/Repositories/CollectionCipherRepository.cs:line 187
   at Bit.Core.Services.CipherService.SaveCollectionsAsync(Cipher cipher, IEnumerable`1 collectionIds, Guid savingUserId, Boolean orgAdmin)
   at Bit.Api.Controllers.CiphersController.PutCollectionsAdmin(String id, CipherCollectionsRequestModel model) in /home/runner/work/server/server/src/Api/Controllers/CiphersController.cs:line 357
   at lambda_method1572(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
</code></pre></details>

I'm also getting some CorsPolicyService errors from the Identity module, but I'm not sure they are related.

**Comment by MGibson1** [comment]
THANK YOU, @holow29 ! Those queries are exactly what we're worried about with the EF offering right now and the reason it's beta. Do you think you can make a separate issue for us to track for it? If not, let me know and I'll make one.

BTW, I have a PR here https://github.com/bitwarden/server/pull/1997 to remove the admin migration issue you mentioned before, I just need to spend some time reproducing and testing to make sure that job is fully excised.

**Comment by lfvjimisola** [comment]
@MGibson1 Are you still working actively on PostgreSQL support? When do you think it would be realistic for it to be released as stable/production? The reason why I am asking is that our organization is considering Bitwarden and we are trying to standardize on PostgreSQL in general.

**Comment by MGibson1** [comment]
@lfvjimisola We are still working on EF (including Postgres) support, yes, but currently see it as an option for self-hosting on low-spec devices.

Honestly, I'm thrilled to hear you're company is looking into Bitwarden and specifically interested in our EF efforts. I'm definitely going to bring this up with our PM team. Better understanding demand for this would help a lot with getting it prioritized higher.

**Comment by lfvjimisola** [comment]
@MGibson1 Thank for the info. To be honest, might consider waiting until the PostgreSQL option is available and from seeing this thread we aren't alone. That said, we are under a time-constraint as we would like to use Bitwarden together with Ansible playbooks. Hope that you'll be able to prioritize it higher. Feel free to reach out if needed.

Is there a plan for a migration tool from MSSQL to PostgreSQL?

**Comment by fsagen** [comment]
> @lfvjimisola We are still working on EF (including Postgres) support, yes, but currently see it as an option for self-hosting on low-spec devices.
> 
> Honestly, I'm thrilled to hear you're company is looking into Bitwarden and specifically interested in our EF efforts. I'm definitely going to bring this up with our PM team. Better understanding demand for this would help a lot with getting it prioritized higher.

I agree with Ifvjimisola, it would be very much appreciated for our system administrators to standardize our database types and not having to keep up with several types like MSSQL, MySQL or MariaDB etc. Other arguments for prioritizing EF support are:

- Most orgs. that run bitwarden self-hosted probably have network segmentation in place and are running the DB externally
- Running a HA cluster for MS SQL is expensive if its only reason for existance is Bitwarden alone
- Some orgs also try to stay away from microsoft in general, i know it's possible to run MSSQL on linux but it's not preferred.

**Comment by kTitan** [comment]
> * Some orgs also try to stay away from microsoft in general, i know it's possible to run MSSQL on linux but it's not preferred.

that’s also the reason why our company has not yet decided to use Bitwarden as our company password safe. 
(ok we are only a small one, which would only need about 20 licenses.)
but I am very sure the some of our customers would also choose to adapt Bitwarden if we have a internal success story about that. Currently I am not able to recommend it for self hosting to them, because of the mssql database. 
Our customers also often do not have the power to decide which database they can use, because there are company policies for that. 

It would be interesting to know, how high the current priority is for this feature. 
if there will be again several month till it becomes stable, I have my doubts that we will be able to wait so long.

**Comment by baurmatt** [comment]
We're a 160 seats customer. We REALLY like to have PostgreSQL production support for Bitwarden! There is just no MSSQL knowledge in our organization to properly handle a MSSQL installation. Instead we have quite extensive knowledge about PSQL/MYSQL.

Is there any timeline when this is planned to be stable?

**Comment by ggiesen** [comment]
I'd like to introduce Bitwarden into our org now that bw-cli has a REST API server, but we need HA and can't justify the expense of MSSQL clustering just for bitwarden (nor do we really want MSSQL at all, we're a postgres/mariadb shop).

Would really love to see PostgreSQL support become production ready.

**Comment by kspearrin** [comment]
Hi everyone. I have exciting developments to report on this front. We've been working on simplifying self hosting of Bitwarden a lot recently. This includes collapsing everything into a single, lightweight docker container, simplifying application configuration, and making it so that you can bring your own database provider (including MySQL/MariaDB and PostgreSQL). More information and official documentation will be announced very soon but wanted to let everyone in this thread know that you can begin tinkering with this new deployment option today and giving us some feedback.

Here's a quick compose file and environment configuration example on how to get started with it:

- Docker image is on Docker Hub under the `dev` tag: https://hub.docker.com/r/bitwarden/self-host
- Compose file: https://github.com/bitwarden/server/blob/master/docker-unified/docker-compose.yml
- Env settings (referenced in the compose file): https://github.com/bitwarden/server/blob/master/docker-unified/settings.env

Note: Consider this a beta. Please do not use this a production environment. It is surely full of bugs and other issues that we need to fix which is why we want your feedback.

**Comment by FingerlessGlov3s** [comment]
@kspearrin when this hits stable/production ready will there be a way to migrate the data out of the current MSSQL container?

**Comment by kspearrin** [comment]
@FingerlessGlov3s That would be a goal, but I am not sure how yet. You could just export your data vault and reimport it back into the new instance, but obviously that has limitations.

**Comment by FingerlessGlov3s** [comment]
> @FingerlessGlov3s That would be a goal, but I am not sure how yet. You could just export your data vault and reimport it back into the new instance, but obviously that has limitations.

I have multiple users, so I want to avoid any disruption if possible. Hopefully just 30-60 minutes down time, which they'll be fine with because Bitwarden can work offline during that period.

**Comment by kspearrin** [comment]
@FingerlessGlov3s Not that with this new way of deploying you can also continue to use MS SQL as a database provider, so your existing database would still port over.

**Comment by lfvjimisola** [comment]
@kspearrin That is great news. FYI, we also need to be able to migrate all data from existing MSSQL to PostgreSQL. Ideally with as little downtime as possible.

**Comment by GieltjE** [comment]
Wow, this is a really good improvement, easy to setup (even when using your own external db server).
Liking the lack of the old cluster of files and the much simpler setup.

The only thing that I couldn't get working is the /admin/ part, just results in a 404.

**Comment by udf2457** [comment]
@kspearrin I've been lurking watching this development because MSSQL dependency is pretty much the only thing putting me off Bitwarden.

However taking a quick look around the repo, is [this file](https://github.com/bitwarden/server/blob/752aa70924f997b9a78644f437df08ad336cc17c/util/PostgresMigrations/Scripts/2021-07-08_00_Init.psql) representative of your proposed Postgres schema ?

Because there's a scary amount of `character varying` that doesn't pass the Postgres smell test.  See the _["Don't Do This"](https://wiki.postgresql.org/wiki/Don't_Do_This)_ wiki on the postgres site.   (e.g. for _varchar(n)_ the TL;DR is you probably should be using _text_ instead).

Also I hope you're going to (eventually ?) be using Postgres stored functions in order to combat against SQL injection attacks ?

**Comment by kspearrin** [comment]
@udf2457 we use Entity Framework Core ORM for the MySQL and PostegreSQL implementations. EF just generates all the SQL.

**Comment by knocte** [comment]
> @udf2457 we use Entity Framework Core ORM for the MySQL and PostegreSQL implementations. EF just generates all the SQL from C#.

I guess this is an answer to the question about stored functions, but what about the schema? I'm guessing EFCore didn't generate the schema.

**Comment by udf2457** [comment]
> > @udf2457 we use Entity Framework Core ORM for the MySQL and PostegreSQL implementations. EF just generates all the SQL from C#.
> 
> I guess this is an answer to the question about stored functions, but what about the schema? I'm guessing EFCore didn't generate the schema.

And even _if_ it did generate the schema, the schema file is not _that_ big that you can't manually tweak it to be better suited for Postgresql instead of blindly using whatever the automated tool spat out.

And _if_ it was automatically generated, then was it reviewed ?  For example is the custom `en-u-ks-primary` collation really necessary ?

**Comment by kspearrin** [comment]
Yes, the schema, queries, and migrations are all generated from Entity Framework Core. All code is peer reviewed, however, we do not have any Postgres expertise on the team, so we mostly rely on what is generated. Any specific feedback will be helpful since I believe tweaks can be made.

**Comment by udf2457** [comment]
I'm a bit busy at $work with various deadlines due to change freezes caused by the upcoming holiday season. So I'm not sure I would be able to commit much time before the new year.

However a quick low-hanging fruit would be as above, i.e.:

1. Any references to `varchar`/`character varying` etc.  changed to `text`, and where there is a genuine need for a length constraint, add `ADD CONSTRAINT _$name_ CHECK (char_length(_$col_) <= _$len_);`
2. `timestamp without timezone` timestamp to `timestamp with timezone`
3. Review the ["Don't Do This"](https://wiki.postgresql.org/wiki/Don't_Do_This) wiki page, it provides a good summary that doesn't require any particular Postgres expertise
4. Have a think about that `en-u-ks-primary` collation ? If it was put there by the automated tool, is it needed ?  Messing around with custom collations on any database should only be done after some thought.

**Comment by kspearrin** [comment]
@udf2457 

Re 1, Can we change these datatypes without affecting data already stored? Seems like we should if we're going from constrained to un constrained lengths?


Re 3, I don't know a lot about database collation, however, I found this in our code that is initing npgsql: https://github.com/bitwarden/server/blob/master/src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs#L107-L109

It references https://www.npgsql.org/efcore/misc/collations-and-case-sensitivity.html?tabs=data-annotations#database-collation and appears we are being specific for some reason. 

Sounds like maybe we have to specify a collation of some sort? I don't know why this specific value (en-u-ks-primary) was chosen.

**Comment by GieltjE** [comment]
Been over the MariaDB/MySQL tables, there are some very good things and some weird things.
1: user.Id is ascii, not UTF (great practice, much better than most other databases out there)
1a: user.[PublicKey|PrivateKey|LicenseKey|ApiKey] are UTF8mb4_general_ci (best choice in my opinion for MySQL/MariaDB for text with possible special characters), which is great for text, but these could be ascii too if we're going for consistency and optimization.
1b: I probably missed a few in this category, guess these apply to more tables/fields (e.g. grant.Key)

2: user.[CreationDate|RevisionDate|LastFailedLoginDate] are named Date but are DateTime

3: [authrequest|cipher|device].UserId is not an index, is this table always queried on Id?
3a: [collection].OrganizationId is not an index, is this table always queried on OrganizationId?
3b: [emergencyaccess|event|folder].Id is the only index, is this table always queried on Id?
3c: stopped looking after the previous ones, these should be easy to find and fix

4: the distinction between char and varchar appears to be correct

5: organization.PlanType is unsigned (correct and good practice), but a lot of other integers are signed while they obviously should be unsigned.

With some minor work and proper care on the index front this is should be a very exciting release.

**Comment by GieltjE** [comment]
@kspearrin Being specific about the collation is a good thing, it shows care being taken when designing the database.

For MariaDB/MySQL there's _ci for Case Insensitive and _bin for hard binary comparisons.

The scheme is generally <encoding>_<locale>_<comparator>, so either utf8mb4_bin (no locale!) or something like utf8mb4_general_ci or utf8mb4_danish_ci. 
Though the country specific ones are rarely used.
Where utf8mb4 appears to have be gaining much traction in the past few years, it just works for nearly all solutions where any form of special characters outside of the ascii charset are used.

**Comment by udf2457** [comment]
@kspearrin 

> Re 1, Can we change these datatypes without affecting data already stored?

I can't think why not, since we're not crossing boundaries like you might be with numeric-type fields. Should be a simple `alter table alter column` command I would have thought.

I guess the only question might be on the indexing side, but a quick drop/re-index would take care of that.

**Comment by GieltjE** [comment]
On any modern DB the indexes should be updated accordingly, did this plenty of times and never manuals touched an index before when doing so.

**Comment by udf2457** [comment]
> On any modern DB the indexes should be updated accordingly, did this plenty of times and never manuals touched an index before when doing so.

Indeed ... I was just trying to think hard for anything that might be affected.  But in reality I think the only thing likely to change during a `varchar`->`text` are background PG catalog things rather than the user-provided data itself.

**Comment by GieltjE** [comment]
Keep in mind that on MariaDB/MySQL [var]char is preferred over [tiny|medium|long|]text types, as the text type cannot be part of indexes on some MySQL versions and they are fixed in size whereas the [var]char type can vary in length.

**Comment by kspearrin** [comment]
Great. Thank you for the feedback. None of these sound too terrible or fundamental and seem like we could address them without issue with schema updates in the future.

**Comment by udf2457** [comment]
Sounds good @kspearrin.

I guess potentially the index definitions for Postgres could be up for review, but then I'm not familiar with the sort of queries you're making.

Certainly if you're doing any of the following, you could do better than a standard index definition:

- Looking for things "anywhere" within a `text`field (i.e. not just prefixes)
- Making queries against `bool` fields where you're only really interested in one of the statuses. 
- Doing anything that might benefit from a partial index or an expression index

**Comment by GieltjE** [comment]
For the idexing checks we could either read the source and check all queries or enable something like general_log on MySQL/MariaDB, perform all possible actions in the UI and read all queries that are now listed in the log file.

Using the EXPLAIN statement should give some more insight (https://mariadb.com/kb/en/explain/).
But for the quick wins just looking at all WHERE ... statements and just using some common sense works fine most of the time.

**Comment by kspearrin** [comment]
There is certainly room for improving indexing in the EF implementations of MySQL and PostgreSQL, however, I think we've got a good understanding of that and we'll work on tuning that over time.

**Comment by kspearrin** [comment]
We have announced an official beta release supporting PostgreSQL and MySQL through our new unified Docker image here: https://bitwarden.com/blog/new-deployment-option-for-self-hosting-bitwarden/

Docs on getting started: https://bitwarden.com/help/install-and-deploy-unified-beta/

Closing this issue now.

**Comment by danpoltawski** [comment]
Is there another issue for tracking migrating out of mssql? (Have a 50+ user deployment I'd really like to migrate onto our existing postgres cluster!)

**Comment by kspearrin** [comment]
@danpoltawski  No, but we are tracking it internally. Feel free to open a public issue.

**Comment by justindbaur** [comment]
Hi All, since you all are interested in using Postgres I just want to give a little visibility to #2480 where we are going to be tracking any issues with Bitwarden unified which includes the path towards using Postgres. If you run into any issue feel free to check there for if it's being tracked already and mention it if you run into something new.

**Comment by holow29** [comment]
Will unified be the only officially supported way of using other DBMS? Would it be wrong to assume that standard self-host architecture has the same support if configured in the config? Haven't dug around the unified release enough yet to see all what it is doing vs. standard.

**Comment by justindbaur** [comment]
@holow29 The standard architecture does have a way of using Postgres & MySQL you have to do exactly what you said configure the config to use it. But note that even using the Standard deployment the Postgres and MySQL options are also a beta.

**Comment by sjmf** [comment]
A note for others who stumble across this issue looking for a way to configure an alternate database server in the [Linux Manual Deployment process](https://bitwarden.com/help/install-on-premise-manual/), as I was.

You will need to add the following settings to `global.env` (or `global.override.env`) – see also [1, 2]:

```
globalSettings__DatabaseProvider=Postgres
globalSettings__PostgreSql__connectionString="Host=postgres;Username=postgres;Password=REPLACE_DEFAULT_PASSWORD;Database=vault"
```

You may also wish to configure an alternate database provider in `docker-compose.yml`, depending on your setup:

```
  postgres:
    image: postgres:16
    container_name: postgres
    restart: always
    volumes:
        - ../postgres:/var/lib/postgresql
    env_file:
        - ../env/pgsql.env
```

I also created a `postgres` folder to contain the data volume.

My `pgsql.env` contains the following:

```
POSTGRES_USER=postgres
POSTGRES_PASSWORD=REPLACE_DEFAULT_PASSWORD
POSTGRES_DB=vault
```

Hope this is helpful. 

Sources:
[1] https://github.com/bitwarden/server/blob/main/.github/workflows/infrastructure-tests.yml
[2] https://github.com/bitwarden/server/blob/main/src/Core/Enums/SupportedDatabaseProviders.cs
